docs: new features in CML 1.6 (#735)
zama-ai/concrete-ml: docs/built-in-models/training.md

# Encrypted training

This document explains how to train [SGD Logistic Regression](../references/api/concrete.ml.sklearn.linear_model.md#class-sgdclassifier) on encrypted data.

> Training on encrypted data is done through an FHE program that is generated by Concrete ML, based on the characteristics of the data that are given to the `fit` function. The FHE program associated to a `SGDClassifier` object, once it is fit on the encrypted data, is specific to the distribution and the dimensionality of that data.

Reviewer: ```suggestion
Training on encrypted data is done through an FHE program that is generated by Concrete ML, based on the characteristics of the data that are given to the `fit` function. Once the FHE program associated with the `SGDClassifier` object has fit the encrypted data, it performs specifically to that data's distribution and dimensionality.
```

> When deploying encrypted training services, developers need to consider the type of data that future users of their services will train on:

Reviewer: ```suggestion
When deploying encrypted training services, you need to consider the type of data that future users of your services will train on:
```

> - the distribution of the data should match to achieve good accuracy
> - the dimensionality of the data needs to match since the deployed FHE programs are compiled for a fixed number of dimensions.

Reviewer: ```suggestion
- The distribution of the data should match to achieve good accuracy
- The dimensionality of the data needs to match since the deployed FHE programs are compiled for a fixed number of dimensions.
```

The trainable logistic model uses Stochastic Gradient Descent (SGD) and quantizes the data, weights, gradients and the error measure. It currently supports training 6-bit models, including g both the coefficients and the bias.

The `SGDClassifier` does not currently support training models with other bit-width values. The execution time to train a model is proportional to the number of features and the number of training examples in the batch. The `SGDClassifier` training does not currently support client/server deployment for training.

## Deployment

> Once you have tested an `SGDClassifier` that trains on encrypted data, is is possible to build an FHE training service by deploying the FHE training program of the `SGDClassifier`. See the [Production Deloyment](../guides/client_server.md) page for more details on how to the Concrete ML deployment utility classes. To deploy an FHE training program, the `mode='training'` parameter must be passed to the `FHEModelDev` class.

Reviewer: ```suggestion
Once you have tested an `SGDClassifier` that trains on encrypted data, you can build an FHE training service by deploying the FHE training program of the `SGDClassifier`. See the [Production Deloyment](../guides/client_server.md) page for more details on how to the Concrete ML deployment utility classes. To deploy an FHE training program, you must pass the `mode='training'` parameter to the `FHEModelDev` class.
```
